developer-docs/architecture/import.md

```diff
@@ -0,0 +1,144 @@
+a technical limitation of the Pulumi engine. In order to actually perform the diff--an
+operation that is required whether or not the user is permitted to desctribe a desired
+state for the imported resource that differs from its actual state--the engine must fetch
+the resource's current imports and state from its provider. In order for this state to
+affect the steps the engine issues for the resources, the state would need to be fetched
+during or prior to the point at which the resource's registration reaches the [step
+generator](resource-registration.md#the-step-generator). In the former case, this would
+cause the engine to spend an unacceptable amount of time in the step generator, as it
+processes resource registrations serially. In the latter case, the user experience would
+likely be negatively affected by a lack of output from the Pulumi CLI, which only displays
+the status of [steps](resource-registration.md#the-step-generator). In order to address
+these issues, the operations described above happen in a dedicated `ImportStep` that is
+run by the [step executor](resource-registration.md#the-step-executor).
+
+![Import diagram](./import.svg)
+
+## `pulumi import`
+
+The second, newer method of importing resources into a stack is the `pulumi import`
+command. This command accepts a list of import specs to import, imports the resources into
+the destination stack, and generates definitions for the resources in the language used
+by the stack's Pulumi program. Each import spec is at least a type token, name, and ID,
+but may also specify a parent URN, provider reference, and package version.
+
+During a `pulumi import`, given a resource R of type T with import ID X and an empty set
+of input properties Iₚ, the engine performs the following sequence of operations:
+
+1. Fetch the current inputs Iₐ and state Sₐ for the resource of type T with ID X from its
+   provider by calling the provider's [`Read` method](providers/implementers-guide.md#read).
+   If the provider does not return a value for Iₐ, the provider does not support importing
+   resources and the import fails.
+2. Fetch the schema for resources of type T from the provider. If the provider is not
+   schematized or if T has no schema, the import fails.
+3. Copy the value of each required input property defined in the schema for T from Iₐ to Iₚ.
+4. Validate the resource's inputs and apply any programmatic defaults by passing Iₚ and Iₐ
+   to the provider's [`Check` method](providers/implementers-guide.md#check). Let Iₖ be the
+   checked inputs; these inputs form the resource's desired state.
+5. Check for differences between Iₖ and Sₐ by calling the provider's [`Diff` method](providers/implementers-guide.md#diff).
+   If the provider reports any differences, the values of the differing properties are
+   copied from Sₐ to Iₚ.
+
+If all of these steps succeed, the user is left with a definition for R in their Pulumi
```

Comment:
Is the definition inserted into the user's Pulumi Program? This kind of implies that it is. `pulumi help import` says that this definition will be printed but that the user should add it.